uefi/src/proto/nvme/mod.rs

```diff
@@ -0,0 +1,285 @@
+/// stay alive until the response was interpreted.
+#[derive(Debug)]
+pub struct NvmeRequest<'buffers> {
+    io_align: u32,
+    cmd: NvmExpressCommand,
+    packet: NvmExpressPassThruCommandPacket,
+    transfer_buffer: Option<AlignedBuffer>,
+    meta_data_buffer: Option<AlignedBuffer>,
+    _phantom: PhantomData<&'buffers u8>,
+}
+
+// NVMe commands consist of a bunch of CDWs (command data words) and a flags bitmask, where
+// one bit per cdw is set when it should be read. Our request builder has one setter method
+// with_cdwX() for every cdw, which also automatically sets the corresponding flag-bit.
+// This macro generates one such setter method.
+macro_rules! define_nvme_command_builder_with_cdw {
+    ($fnname:ident: $fieldname:ident => $flagmask:expr) => {
+        /// Set the $fieldname parameter on the constructed nvme command.
+        /// This also automatically flags the parameter as valid in the command's `flags` field.
+        ///
+        /// # About NVMe commands
+        /// NVMe commands are constructed of a bunch of numbered CDWs (command data words) and a `flags` field.
+        /// The `flags´ field tells the NVMe controller which CDWs was set and whether it should respect
+        /// the corresponding CDWs value.
+        /// CDWs have no fixed interpretation - the interpretation depends on the command to execute.
+        /// Which CDWs have to be supplied (and enabled in the `flags` field) depends on the command that
+        /// should be sent to and executed by the controller.
+        /// See: <https://nvmexpress.org/specifications/>
+        #[must_use]
+        pub const fn $fnname(mut self, $fieldname: u32) -> Self {
+            self.req.cmd.$fieldname = $fieldname;
+            self.req.cmd.flags |= $flagmask.bits();
+            self
+        }
+    };
+}
+
+/// Builder for constructing an NVMe request.
+///
+/// This structure provides convenient methods for configuring NVMe commands,
+/// including parameters like command-specific data words (CDWs)
+/// and optional buffers for transfer and metadata operations.
+///
+/// It ensures safe and ergonomic setup of NVMe requests.
+///
+/// # Lifetime
+/// `'buffers`: Makes sure the io-buffers bound to the built request
+/// stay alive until the response was interpreted.
+#[derive(Debug)]
+pub struct NvmeRequestBuilder<'buffers> {
+    req: NvmeRequest<'buffers>,
+}
+impl<'buffers> NvmeRequestBuilder<'buffers> {
+    /// Creates a new builder for configuring an NVMe request.
+    ///
+    /// # Parameters
+    /// - `io_align`: Memory alignment requirements for buffers.
+    /// - `opcode`: The opcode for the NVMe command.
+    /// - `queue_type`: Specifies the type of queue the command should be placed into.
+    ///
+    /// # Returns
+    /// An instance of [`NvmeRequestBuilder`] for further configuration.
+    #[must_use]
+    pub fn new(io_align: u32, opcode: u8, queue_type: NvmeQueueType) -> Self {
```

Comment:
I think we could add an assert to check that `io_align: u32` is a power of 2, but that is not urgent to get this merged